### DISCUSSION TITLE
Implementing SSE Using Extended JS API

### DISCUSSION THREAD

**Discussion by naroop** [pocketbase/pocketbase#3195] [Q&A] [answered]
Hello, I am attempting to create an application that needs SSE to send out notifications. The following code snippet is a rudimentary version what I have so far:
```javascript
// main.pb.js

const clients = []

routerAdd('GET', '/connection', (c) => {
  c.response().header().set('Content-Type', 'text/event-stream; charset=UTF-8');
  c.response().header().set('Cache-Control', 'no-store');
  c.response().header().set('Connection', 'keep-alive');
  
  c.response().write('data: Connected.\n\n')
  clients.push(c.response())
});

routerAdd('GET', '/send-message', (c) => {
  clients.forEach((client) => {
    client.write('data: Message sent!\n\n');
  });
});
```
```typescript
// vue.js Script
...
const eventSource = new EventSource('http://localhost:8090/connection');

eventSource.onopen = () => {
  sseStatus.value = 'Connection created';
};
eventSource.onmessage = (event) => {
  sseData.value = event.data;
};
eventSource.onerror = (error) => {
  console.error('Error:', error);
};
...
```
I have been messing around with it for a few hours now. Sending a request to this URL from the Vue application I made will just result in it getting the EventStream message but then repeatedly reconnecting every couple seconds from that point on. It is not persisting the connection. I have duplicated this exact example in an express.js API and it works just fine, so I don't know if I am missing something that is PocketBase/JS SDK specific.

It seems like after receiving the message the frontend just immediately disconnects.

I have tried looking through the docs to see how it was accomplished in Go to create the collection subscriptions but can't seem to find anything that helps with the Javascript. Also tried extending the API with Go instead but I don't know the language really at all and couldn't figure out anything different that would help. 

Any help or suggestions at all would be much appreciated!

**Comment by ganigeorgiev** [comment]
This will be difficult to be accomplished with the JS app hooks.

First, the route handlers can't access top/root level defined variables (see https://pocketbase.io/docs/js-overview/#handlers-scope). If you want to share variables across multiple handlers, you can store them in `$app.cache().set("customClients", clients)` (to retrieve call `$app.cache().get("customClients")`).

In order to establish SSE connection you'll need some sort of infinite loop (eg. `while(1) {}`) in the handler to keep the connection open. To send messages you'll need to call `c.response().write(line)` and `c.response().flush()` at the end.

Note that the execution in the js app hook handler is synchronous and broadcasting a message to multiple clients in a loop will block on each iteration until the client has received the message.

But even you manage to make the above works, the biggest problem will be dealing with races (eg. a client connection has been closed while you are iterating in the broadcasting loop).

In general if you need more control over the execution flow, I'd recommend using the Go APIs instead since you can leverage better the builtin concurrency primitives.

Could you elaborate a little more on your use case?

**Reply by ganigeorgiev** [reply, marked as answer]
This of course will depend on your actual use case, but I think the easiest way to handle notification subscriptions is to use the existing realtime apis.

1. Create a base `notifications` collection with a "data" `json` field (don't forget to setup the List and View API rules).

2. To submit new notification, you can just call client-side `pb.collection("notifications").create({ data: {something: 123} })`

3. To subscribe to the `notifications` changes, you can use client-side `pb.collection("notifications").subscribe("*", (e) => { ... })`

This has several benefits:
  - you can add custom structured fields to each message (eg. flag to indicated "read/seen")
  - you can maintain history of the messages or define a retention policy, to prevent the table growing too big; for the latter you can schedule a [cron job](https://pocketbase.io/docs/js-jobs-scheduling/) that will cleanup older records, something like (_I haven't tested it_):
    ```js
    // pb_hooks/main.pb.js
    // runs every day at midnight
    cronAdd("cleanup_notifications", "0 0 * * *", () => {
        const date = new DateTime()

        // change the now datetime to yesterday
        date.time().addDate(0, 0, -1)

        $app.dao().db()
            .delete("notifications", $dbx.exp("created < {:date}", {
                date: date.string(),
            }))
            .execute()
    })
    ```
   - you don't have to deal manually with the various concurrent edge cases that could arise during the lifetime of the SSE connection

**Reply by naroop** [reply]
Thanks for the detailed and thought out answer. 

I am working on what is essentially a simplified version of VSCode live-share in the browser, to be used for educational or demonstrative purposes. A host can start a session and have a bunch of students join using a room code and be able to view/edit a code editor and console.

This notification system would be a replacement for a WebSocket (i wasn't sure websocket would work here at all) in that it would manage sending out data to all connected clients in a room when needed.

A sample use case would be the host gives Student 1 edit privileges, which sends out a notification to all connected users that Student 1 is now editing. Or I thought of a hand-raise feature where a student could signal a hand raise, and all connected users would receive this notification, and then show some kind of hand icon near that user.

Given what you explained in your first response, I believe that your suggestion of using a notifications collection would probably be the best solution here. I will try that out sometime soon. Thanks again for responding.
